subsys/bluetooth/gatt_pool.c

```diff
@@ -235,121 +266,188 @@ static void uuid_unregister(struct bt_uuid const *uuid)
-	svc->attr_count++;
-}
-
-void bt_gatt_pool_chrc_put(struct bt_gatt_attr const *attr)
-{
-	__ASSERT_NO_MSG(attr != NULL);
+	ret = uuid_register((struct bt_uuid **) &dest_chrc->uuid, chrc->uuid);
+	if (ret) {
+		return ret;
+	}
 
-	uuid_unregister(((struct bt_gatt_chrc *) attr->user_data)->uuid);
-	chrc_release(attr->user_data);
-	uuid_unregister(attr->uuid);
+	gp->svc.attr_count++;
+	return 0;
 }
 
-void bt_gatt_pool_desc_get(struct bt_gatt_service *svc,
-			   struct bt_gatt_attr const *descriptor)
+int bt_gatt_pool_desc_alloc(struct bt_gatt_pool *gp,
+			    struct bt_gatt_attr const *descriptor)
 {
-	struct bt_gatt_attr *attr = &svc->attrs[svc->attr_count];
+	int ret;
+	struct bt_gatt_attr *attr = &gp->svc.attrs[gp->svc.attr_count];
 
-	__ASSERT_NO_MSG(svc->attrs != NULL);
+	if (!gp || !gp->svc.attrs || !descriptor) {
+		LOG_ERR("Invalid attribute");
+		return -EINVAL;
+	}
+	if (gp->svc.attr_count >= gp->attr_array_size) {
+		LOG_ERR("No space left on given svc");
+		return -ENOSPC;
+	}
 
 	memset(attr, 0, sizeof(*attr));
 
 	memcpy(attr, descriptor, sizeof(*attr));
 	attr->uuid = NULL;
-	uuid_register((struct bt_uuid **) &attr->uuid, descriptor->uuid);
-
-	__ASSERT_NO_MSG(svc->attr_count < CONFIG_BT_GATT_HIDS_ATTR_MAX);
-	svc->attr_count++;
-}
-
-void bt_gatt_pool_desc_put(struct bt_gatt_attr const *attr)
-{
-	__ASSERT_NO_MSG(attr != NULL);
+	ret = uuid_register((struct bt_uuid **) &attr->uuid, descriptor->uuid);
+	if (ret) {
+		return ret;
+	}
 
-	uuid_unregister(attr->uuid);
+	gp->svc.attr_count++;
+	return 0;
 }
 
-void bt_gatt_pool_ccc_get(struct bt_gatt_service *svc,
-			  struct _bt_gatt_ccc const *ccc)
+int bt_gatt_pool_ccc_alloc(struct bt_gatt_pool *gp,
+			   struct _bt_gatt_ccc const *ccc)
 {
-	struct bt_gatt_attr *attr = &svc->attrs[svc->attr_count];
+	int ret;
+	struct bt_gatt_attr *attr = &gp->svc.attrs[gp->svc.attr_count];
 	struct bt_uuid      *uuid_gatt_ccc = BT_UUID_GATT_CCC;
 
-	__ASSERT_NO_MSG(svc->attrs != NULL);
+	if (!gp || !gp->svc.attrs || !ccc) {
+		LOG_ERR("Invalid attribute");
+		return -EINVAL;
+	}
+	if (gp->svc.attr_count >= gp->attr_array_size) {
+		LOG_ERR("No space left on given svc");
+		return -ENOSPC;
+	}
 
 	memset(attr, 0, sizeof(*attr));
 
-	uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_ccc);
+	ret = uuid_register((struct bt_uuid **) &attr->uuid, uuid_gatt_ccc);
+	if (ret) {
+		return ret;
+	}
 	attr->perm = BT_GATT_PERM_READ | BT_GATT_PERM_WRITE;
 	attr->read = bt_gatt_attr_read_ccc;
 	attr->write = bt_gatt_attr_write_ccc;
-	ccc_get((struct _bt_gatt_ccc **) &attr->user_data);
+	ret = ccc_get((struct _bt_gatt_ccc **) &attr->user_data);
+	if (ret) {
+		return ret;
+	}
 	memcpy(attr->user_data, ccc, sizeof(struct _bt_gatt_ccc));
 
-	__ASSERT_NO_MSG(svc->attr_count < CONFIG_BT_GATT_HIDS_ATTR_MAX);
-	svc->attr_count++;
+	gp->svc.attr_count++;
+	return 0;
 }
 
-void bt_gatt_pool_ccc_put(struct bt_gatt_attr const *attr)
+
+void bt_gatt_pool_attr_free(struct bt_gatt_attr const *attr)
```

Comment:
This function will free the from the UUID/CHRC/CC pools. But will not free the attribute itself (gp->svc.count is untouched). Which makes sense. So I think we should make this function static. Because it is not really useful on it's own, only when called by bt_gatt_pool_free.

Reply:
Changed to static